build-tools/automation/yaml-templates/run-sliced-nunit-tests.yaml

```diff
@@ -2,22 +2,55 @@ parameters:
   testAssembly:           # NUnit test assembly to run
   testFilter:             # Filter used to select tests (NUnit test selection language, not dotnet test filter language)
   testRunTitle:           # Title of the test run
-  testResultsTitle:       # Title used to construct test results file name
+  retryFailedTests: true  # Retry failed tests once
```

Comment:
General thought but might it be better to add the [Retry] attribute to tests that we see failing somewhat regularly? NUnit does provide _some_ built in support for this that we are already using in a handful of places.

Reply:
While I agree we should be trying to improve our test reliability, it does not seem like something we have been dedicating time to or are likely to start dedicating time to.

We have this data today: https://devdiv.visualstudio.com/DevDiv/_test/analytics?definitionId=11410&contextType=build, but it doesn't appear to be localized to a few tests.  (550 unique tests failed over the past 14 days.)  However, this may point to an issue with the suite(s) themselves that could be investigated.

It is good to point out that this data will not be available after this PR (unless the test fails on retry), as we will no longer be reporting initial failures to AzDO so that we have a green build.  If we would like to continue having this data, a good compromise might be to perform automatic retries on PR builds but not for `main` builds.

I feel like the large number of unique test failures makes `[Retry]` a poor option.  I looked at applying it at the class or assembly level, but that isn't supported. It can only be placed on individual tests.

Grendel and I also discovered that `[Retry]` is not available in `NUnitLite` which is used by the on-device tests.

Reply:
Breaking this down a bit further you can sort by test file, and you can see the vast majority are coming from Designer tests (though I'm not sure what branch). Other than that, we clearly have a few core culprits under a ~90% pass rate:

![image](https://user-images.githubusercontent.com/2000163/235780828-3b15482c-5028-43eb-9d96-7c9352e3bbb2.png)

![image](https://user-images.githubusercontent.com/2000163/235780648-44aee8ec-c3a2-43ad-bc35-161ac37517fb.png)

Maybe it would be best to only enable this for the emulator test jobs for now, as that is where most issues seem to be? This seems like a good interim solution, at least until we can find a machine pool that will provide a more stable emulator/device for us.

I believe most of our test related CI headaches stem from using an emulator in nested virtualization, and could be resolved by investigating different test execution environments.

Reply:
👍 Set `retryFailedTests: false` for the regular MSBuild job, so it will only run on emulator tests.

Reply:
> I believe most of our test related CI headaches stem from using an emulator in nested virtualization, and could be resolved by investigating different test execution environments.

I also wonder if we've ever considered we could have a bug in our code but we just blame the emulator rather than investigating.  :thinking: